docs/proposals/odo-deploy.md

```diff
@@ -0,0 +1,152 @@
+
+`odo deploy delete` could provide developers with a way to clean up any existing deployment of the application. 
+
+### High-level design:
+
+#### Technical considerations:
+- The implementation would be under the experimental flag.
+- Only supported for Devfile v2.0.0 components.
+- Only supported for Kubernetes/ OpenShift targets.
+- `odo deploy` will not make any changes to the user's project.
+
+#### odo deploy 
+This command will build a container image for the application and deploy it on the target Kubernetes environment. 
+
+Flags:
+ - `--tag`: The tag to be used for the built application container image - `<registry>/<org>/<name>:<tag>` (mandatory).
+ - `service-account`: The service account for running privileged containers and push access to the image registry.
+ - `--credentials`: The credentials needed to push the image to the container image registry (mandatory).
+
+#### odo deploy delete
+This command will delete any resources created by odo deploy.
+
+### Detailed design: 
+
+### Devfile
+
+For the initial implementation, we could use devfile v2.0.0 and capture basic outer-loop information as `attributes` on the devfile. `odo` could look for  specific keys, while other tools like Che could ignore them.
+
+For example: 
+```
+attributes:
+    build-dockerfile: <URI>
+    deployment-manifest: <URI>
+```
+
+This should not need any change to the devfile v2 schema or the parser.
+
+### Dockerfile
+This could be any valid dockerfile.
+
+### Deployment Manifest
+The deployment manifest could be templated to help with replacing key bits of information:
+- PROJECT_NAME
+- CONTAINER_IMAGE
+- PORT
+
+For example: 
+```
+apiVersion: app.stacks/v1beta1
+kind: RuntimeComponent
+metadata:
+    name: PROJECT_NAME
+spec:
+    applicationImage: CONTAINER_IMAGE 
+    service:
+        type: ClusterIP
+        port: PORT
+    expose: true
+    storage:
+        size: 2Gi
+        mountPath: "/logs"
+```
+
+### odo deploy
+This command will perform the following actions:
+
+#### Input validation
+- Check if the devfile  is v2.0.0 and that it specifies the expected outer-loop attributes. 
+    - If not provided, display a meaningful error message.
+- Validate all arguments passed by the user. 
+    - If argument values are invalid, display a meaningful error message.
+
+#### Build
+- Start a new pod with a build container (privileged container with `buildah:stable` image).
+- Mount the image registry credentials passed by the user.
+- Copy the application source code into the build container.
+- Fetch the dockerfile using URI in the attributes of the devfile and make it available in the container.
+- Execute `buildah bud` command using src code, dockerfile and tag information.
+- Push the built image using the tag and credentials.
+- Delete the build container and the pod.
+
+#### Deploy
+- Delete any existing deployments (if the user runs `odo deploy` multiple times).
```

Comment:
Probably an implementation level point, but I think the behaviour should be similar to what `odo push` and `odo push -f` do. Presently, you need to do `odo push -f` if you'd like to delete the component pushed by an earlier run of `odo push`?

Reply:
Agree. It seems like a reasonable thing to do and keep them both consistent. I'll update the proposal with this.

Reply:
This is how `odo push` works currently:
```sh
$ odo push                                                                                                                                                                                    
                                                                                                                                                                                              
Validation                                                                                                                                                                                    
 ✓  Validating the devfile [24381ns]                                                                                                                                                          
                                                                                                                                                                                              
Creating Kubernetes resources for component nodejs                                                                                                                                            
 ✓  Waiting for component to start [1ms]                                                                                                                                                      
                                                                                                                                                                                              
Applying URL changes                                                                                                                                                                          
 ✓  URLs are synced with the cluster, no changes are required.                                                                                                                                
                                                                                                                                                                                              
Syncing to component nodejs                                                                                                                                                                   
 ✓  Checking file changes for pushing [556622ns]                                                                                                                                              
 ✓  Syncing files to the component [209ms]                                                                                                                                                    
                                                                                                                                                                                              
Executing devfile commands for component nodejs                                                                                                                                               
 ✓  Executing devBuild command "npm install" [2s]                                                                                                                                             
 ✓  Executing devRun command "nodemon app.js" [1s]                                                                                                                                            
                                                                                                                                                                                              
Pushing devfile component nodejs                                                                                                                                                              
 ✓  Changes successfully pushed to component                                                                                                                                                  
                                                                                                                                                                                              
$ odo push

Validation
 ✓  Validating the devfile [21294ns]

Creating Kubernetes resources for component nodejs
 ✓  Waiting for component to start [5ms]

Applying URL changes
 ✓  URLs are synced with the cluster, no changes are required.

Syncing to component nodejs
 ✓  Checking file changes for pushing [258076ns]
 ✓  No file changes detected, skipping build. Use the '-f' flag to force the build.

Pushing devfile component nodejs
 ✓  Changes successfully pushed to component
```

Key part being:
```
Syncing to component nodejs
 ✓  Checking file changes for pushing [258076ns]
 ✓  No file changes detected, skipping build. Use the '-f' flag to force the build.
```
Now, I'm not sure if `odo deploy`'s going to, or even required to, check for any file changes between two runs of `odo deploy`. As you pointed, keeping behaviour consistent would be helpful, IMO, if file changes matter to `odo deploy` as they currently do to `odo push`.

Reply:
Have updated the proposal to just create/update the deployment by default.. and when used with -f, it will delete and then re-create.